### DISCUSSION TITLE
Netbox 4.0 compatibility

### DISCUSSION THREAD

**Discussion by yaiqsa** [netbox-community/ansible_modules#1227] [Q&A]
Hi everyone,

I might be looking in all the wrong places, but I can't find any reference to the recent Netbox 4.0 release in this repo & docs.

So I would like to ask a few questions:

- Is this project is already compatible with Netbox 4.0?
- If so, what are your experiences with this collection with Netbox 4.0 instances? (especially `nb_inventory`)
- Noob question: _which_ API (REST or GraphQL) does `nb_inventory` use, and will the new features for that API be used to improve the gathering speed?
  - Because the GraphQL engine was replaced, the strawberry-django-plus query optimizer might prove usefull?
  - Or if REST is used, the `fields=` parameter might speed up queries that return more than necessary at this moment?
  
Thank you in advance!

**Comment by aopdal** [comment]
You can use this on NetBox v4.0+. The inventory is usig the REST API.

**Reply by yaiqsa** [reply]
Actually, a problem was found:

https://github.com/netbox-community/ansible_modules/issues/1265

pynetbox had an issue where they didn't pass the auth token in the first request to Netbox (to /api/). Before 4.0 that page was free for anyone to lookup, but since 4.0 you get a 403 when you request it without authentication.

pytnetbox fixed it in v7.3.4, so when ansible uses that version, it should work again 🥳 

For my installation (Ubuntu 22.04) `pip3 install pynetbox==7.3.4` did the trick

**Comment by chatasos** [comment]
Hi, 

Do you think it would be possible to integrate a combatibility check, just before running the actual tasks?

**Reply by yaiqsa** [reply]
Hmm. Depends on what you're looking for I guess..

- Technically it sounds pretty easy to implement a check in the Ansible module where it checks if the Netbox server version is between version X and version Y. However, that means that once a new Netbox version comes out, this version has to be tested somewhere, and added to the check. If that doesn't happen, it will just fail while there might not actually be a problem.
  - Additionally, as you see in this thread, it might not be known that there are incompatibilities (in dependencies), so such a 'static' check isn't of much value.
  - If you take a look at the [pynetbox README](https://github.com/netbox-community/pynetbox) you see a short compatibility list. If you would enforce this, you couldn't use Netbox versions higher than 4.0.6 at the moment. However, I run pynetbox 7.3.4 with Netbox 4.1.2, without any issues..
  - I don't know how often the Ansible modules or pynetbox break backwards compatibility, but if that happens a check might indeed be useful. Then again, if one feature is broken, would you want a compatibility error in all other usecases as well?
- If it is worthwhile for your environment you could always deploy a testing environment yourself. The best way to find out if everything is compatible, is by running your tasks 😅 . So you could upgrade Netbox on your testing environment first, run your playbooks on test systems, and check if problems occur. Of course, this is a quite expensive solution.
- If you want to be stable without too much hassle, I'd recommend just lagging behind on Netbox updates. That way other people will have stumbled across compatibility issues first, and have probably fixed them by the time you upgrade.

And best of all; any compatibility check on the API would have been broken by the upgrade to Netbox 4.0, because the whole problem was that the API 'root' page had a breaking change by requiring authentication.

**Reply by sc68cal** [reply]
> if the Netbox server version is between version X and version Y

I'll just chime in here that we do this already in some parts of the code where there were breaking changes in the NetBox API, that we had to deal with because of our version compatibility policy.
